[GTK] Try to reuse the GTK EGL display instead of always creating a new one

carlosgcampos opened the pull request:
#### 53dd1f1cd48dffd5b1f67470714e7c389a1795bb
<pre>
[GTK] Try to reuse the GTK EGL display instead of always creating a new one
<a href="https://bugs.webkit.org/show_bug.cgi?id=255721">https://bugs.webkit.org/show_bug.cgi?id=255721</a>

Reviewed by Adrian Perez de Castro.

* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::terminateEGLDisplay):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
(WebCore::PlatformDisplay::gtkEGLDisplay):
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp:
(WebCore::PlatformDisplayWayland::initialize):
(WebCore::PlatformDisplayWayland::gtkEGLDisplay):
(WebCore::PlatformDisplayWayland::initializeEGLDisplay):
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.h:
* Source/WebCore/platform/graphics/x11/PlatformDisplayX11.cpp:
(WebCore::PlatformDisplayX11::gtkEGLDisplay):
(WebCore::PlatformDisplayX11::initializeEGLDisplay):
* Source/WebCore/platform/graphics/x11/PlatformDisplayX11.h:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::createSource):
(WebKit::AcceleratedBackingStoreDMABuf::makeContextCurrent):
(WebKit::gtkGLContextIsEGL): Deleted.

Canonical link: <a href="https://commits.webkit.org/263306@main">https://commits.webkit.org/263306@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e2a2edf0892e78f9fb28721224799249c74af3b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4103 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4215 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5449 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4255 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3984 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4194 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4078 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4200 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4057 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3615 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5380 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1745 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3591 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5705 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3570 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3651 "3 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5104 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4058 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3265 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3573 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3591 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3625 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/484 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3853 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp:
```diff
@@ -122,6 +123,43 @@ void PlatformDisplayWayland::initialize()
+        gdk_gl_context_make_current(context.get());
+        m_eglDisplay = eglGetCurrentDisplay();
+    }
+#endif
```
This leaks the window returned by `gtk_window_new()`, it would be better to adopt the pointer into a `GRefPtr`:

 ```suggestion
    auto window = adoptGRef(gtk_window_new(GTK_WINDOW_POPUP));
    gtk_widget_realize(window.get());
    if (auto context = adoptGRef(gdk_window_create_gl_context(gtk_widget_get_window(window.get()), nullptr))) {
        gdk_gl_context_make_current(context.get());
        m_eglDisplay = eglGetCurrentDisplay();
    }
#endif
```

carlosgcampos replied:
Right, I forgot the destroy, but widgets are initially unowned so you can't adopt the floating ref. I'll just use destroy. Thanks!

aperezdc replied:
Adopting uses `g_object_ref_sink()`, a smart pointer would work. If you prefer to use `gtk_widget_destroy()` that's fine as well.

carlosgcampos replied:
No, adopting doesn't do any ref. For widgets it's preferred to use destroy.

aperezdc reviewed (changes requested):
Patch looks reasonable to me, modulo one `GtkWindow` leak so please take a  look at the comment.

carlosgcampos committed:
[GTK] Try to reuse the GTK EGL display instead of always creating a new one
https://bugs.webkit.org/show_bug.cgi?id=255721

Reviewed by Adrian Perez de Castro.

* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::terminateEGLDisplay):
* Source/WebCore/platform/graphics/PlatformDisplay.h:
(WebCore::PlatformDisplay::gtkEGLDisplay):
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp:
(WebCore::PlatformDisplayWayland::initialize):
(WebCore::PlatformDisplayWayland::gtkEGLDisplay):
(WebCore::PlatformDisplayWayland::initializeEGLDisplay):
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.h:
* Source/WebCore/platform/graphics/x11/PlatformDisplayX11.cpp:
(WebCore::PlatformDisplayX11::gtkEGLDisplay):
(WebCore::PlatformDisplayX11::initializeEGLDisplay):
* Source/WebCore/platform/graphics/x11/PlatformDisplayX11.h:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::createSource):
(WebKit::AcceleratedBackingStoreDMABuf::makeContextCurrent):
(WebKit::gtkGLContextIsEGL): Deleted.

Canonical link: https://commits.webkit.org/263306@main
